`AsnType` derive fails to compile for generic enums

reproduction on the current (0.22.0) version:
```rust
use rasn::{types::OctetString, AsnType};

#[derive(AsnType)]
#[rasn(choice)]
pub enum Example<T> {
    String(OctetString),
    Sequence(Vec<T>),
}
```
produces the errors
```
error[E0401]: can't use generic parameters from outer item
 --> src/lib.rs:7:18
  |
5 | pub enum Example<T> {
  |                  - type parameter from outer item
6 |     String(OctetString),
7 |     Sequence(Vec<T>),
  |                  ^ use of generic parameter from outer item
  |
  = note: a `const` is a separate item from the item that contains it

error[E0277]: the trait bound `T: AsnType` is not satisfied
 --> src/lib.rs:7:14
  |
7 |     Sequence(Vec<T>),
  |              ^^^^^^ the trait `AsnType` is not implemented for `T`, which is required by `Vec<T>: AsnType`
  |
  = note: required for `Vec<T>` to implement `AsnType`
help: consider restricting type parameter `T`
  |
5 | pub enum Example<T: rasn::AsnType> {
  |                   +++++++++++++++
```

looks like `TAG_TREE` needs to use a `const` block which will allow it to access the generic parameter, and also needs to add a bound for `T` by default to bound it for `AsnType`. I'll work up a PR to fix these.

--- a/macros/macros_impl/src/asn_type.rs
+++ b/macros/macros_impl/src/asn_type.rs
@@ -62,7 +62,7 @@ pub fn derive_struct_impl(
                 paren_token: Some(<_>::default()),
                 modifier: syn::TraitBoundModifier::None,
                 lifetimes: None,
-                path: syn::parse_str("rasn::AsnType").unwrap(),
+                path: syn::parse_quote!(#crate_root::AsnType),
             }));
     }
 

--- a/macros/macros_impl/src/enum.rs
+++ b/macros/macros_impl/src/enum.rs
@@ -39,10 +39,10 @@ impl Enum {
 
             quote! {
                 {
-                    const VARIANT_LIST: &'static [#crate_root::types::TagTree] = &[#(#field_tags),*];
-                    const VARIANT_TAG_TREE: #crate_root::types::TagTree = #crate_root::types::TagTree::Choice(VARIANT_LIST);
-                    const _: () = assert!(VARIANT_TAG_TREE.is_unique(), #error_message);
-                    VARIANT_TAG_TREE
+                    let variant_list: &'static [#crate_root::types::TagTree] = const { &[#(#field_tags),*] };
+                    let variant_tag_tree: #crate_root::types::TagTree = #crate_root::types::TagTree::Choice(variant_list);
+                    assert!(variant_tag_tree.is_unique(), #error_message);
+                    variant_tag_tree
                 }
             }
         } else {
@@ -50,13 +50,24 @@ impl Enum {
         };
 
         let name = &self.name;
-        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();
+        let mut generics = self.generics.clone();
+
+        for param in &mut generics.params {
+            if let syn::GenericParam::Type(type_param) = param {
+                type_param
+                    .bounds
+                    .push(syn::parse_quote!(#crate_root::AsnType));
+            }
+        }
+
+        let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
         let return_val = self
             .config
             .tag
             .is_some()
             .then(|| quote!(#crate_root::types::TagTree::Leaf(Self::TAG)))
-            .unwrap_or_else(|| quote!(TAG_TREE));
+            .unwrap_or_else(|| quote!(tag_tree));
 
         let (base_variants, extended_variants): (Vec<_>, Vec<_>) = self
             .variants
@@ -174,10 +185,12 @@ impl Enum {
                 const TAG: #crate_root::types::Tag = {
                     #tag
                 };
-                const TAG_TREE: #crate_root::types::TagTree = {
-                    const LIST: &'static [#crate_root::types::TagTree] = &[#tag_tree];
-                    const TAG_TREE: #crate_root::types::TagTree = #crate_root::types::TagTree::Choice(LIST);
-                    const _: () = assert!(TAG_TREE.is_unique(), #error_message);
+                const TAG_TREE: #crate_root::types::TagTree = const {
+                    let list: &'static [#crate_root::types::TagTree] = const {
+                        &[#tag_tree]
+                    };
+                    let tag_tree: #crate_root::types::TagTree = #crate_root::types::TagTree::Choice(list);
+                    assert!(tag_tree.is_unique(), #error_message);
                     #return_val
                 };
                 #alt_identifier

--- a/tests/derive.rs
+++ b/tests/derive.rs
@@ -144,6 +144,13 @@ pub enum ExplicitChoice {
     ByKey,
 }
 
+#[derive(AsnType, Decode, Encode)]
+#[rasn(choice)]
+pub enum GenericEnum<T: Clone, const N: usize> {
+    FixedString(FixedOctetString<N>),
+    Sequence(Vec<T>),
+}
+
 #[derive(AsnType, Clone, Debug, Decode, Encode, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct BasicConstraints {
     #[rasn(default)]

